### PR TITLE
BIP 127: Simple Proof-of-Reserves Transactions

### PR DESCRIPTION
**Pull request by stevenroose**
(submitted to bitcoin-dev, awaiting approval)

### CODE REVIEW

**Comment by luke-jr** [comment]
This needs a backward compatibility section (probably simply saying it's n/a).

Also, @achow101 needs to ACK the modifications to BIP 174.

**Comment by achow101** [comment]
You should add the new type to BIP 174's [Appendix A](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#appendix-a-data-types-and-their-specifications).  That should also include a link to the location in the new BIP descriibing it.

**Comment by stevenroose** [comment]
@achow101 I added the field to that table. I also mention the BIP in the explanation, but it's currently set to `TBD` until this BIP gets a number assigned, then I can update both mentions of TBD to become a hyperlink.

@luke-jr added compatibility section.

**Comment by luke-jr** [comment]
Assigned BIP 127

**Comment by stevenroose** [comment]
I just realized it might be useful to have a way to deal with funds that are locked in a CLTV in the future. Are their opinions about this?

I'm thinking of two approaches:
1. have a global "locktime" variable for the entire proof, which proves that "the prover would be able to spend C coins at time L"
2. allow specifying a locktime for some UTXOs which allows for more fine-grained locktime limitation specification, but might make the "proof statement" more complex. Something like "the prover has ownership of C1 coins now and C2 coins a time L1, ...". For UX purposes the prover can decide to use the same locktime for all it's CLTV UTXOs even though the locktime is further away for some of the UTXOs. Just to simplify the statement.

**Comment by stevenroose** [comment]
@luke-jr I updated the files to use the assigned number. Up to you if you merge it like this and allow me to update once I figured out a good way to deal with CLTV or if you want to wait for that in order to merge this.

**Comment by luke-jr** [comment]
> "the prover would be able to spend C coins at time L"

This can't be proven. The TXOs might get spent before time L.

**Comment by luke-jr** [comment]
```
First Comments-URI must be exactly "https://github.com/bitcoin/bips/wiki/Comments:BIP-0127" in bip-0127.mediawiki
```

**Comment by achow101** [comment]
ACK BIP174 changes

**Comment by stevenroose** [comment]
@luke-jr There is an error in the Travis that I don't understand. Seems to be related to the git history. I tried rebasing from latest master but that didn't work..

**Comment by stevenroose** [comment]
@luke-jr ping :)

**Comment by luke-jr** [comment]
It's saying you need to update the README

File: bip-0127.mediawiki [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+Our specification consists of two parts:
+# the format for the actual proofs
+# a file format used to package a set of proofs and relevant metadata
+
+The final construction should have the following properties:
+* flexible proof construction to support complex wallet infrastructures
+* easy integration with existing wallet solutions (both hardware and software wallets)
+* support for verification via a standard procedure, regardless of publisher of the proof
+* proof prevents reuse of proofs by other parties by commiting to a message
```
Comment (rex4539) [inline bip-0127.mediawiki]: Typo: "commiting"

File: bip-0127.mediawiki [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+# constructing and combining multiple proofs
+#:Having thousands of UTXOs spread across different offline and online wallets could make it difficult to construct a single proof transaction with all UTXOs.  Allowing multiple proof transactions with the same commitment message and block number gives extra flexibility to custodians with complex wallet infrastructure without making the combined proof less secure.
+# metadata for verification
+#:Not all systems that will be used for verification have access to a full index of all transactions.  However, proofs should be easily verifiable even after some of the UTXOs used in the proof are no longer unspent.  Metadata present in the proof allows for relatively efficient verification of proofs even if no transaction index is available.
+# potential future improvements
+#:The extensible metadata format allows for amending the standard in the future.  One potential improvement would be having UTXO set commitments.  These would allow the proofs-of-reserves to come with accompanying proofs-of-inclusion of all used UTXOs in the UTXO set at the block of proof constsruction (making validation even more efficient).
```
Comment (rex4539) [inline bip-0127.mediawiki]: Typo: "constsruction"

File: bip-0127.mediawiki (hunk 1 above)
Comment (stevenroose) [reply bip-0127.mediawiki]: done, thanks

File: bip-0127.mediawiki (hunk 2 above)
Comment (stevenroose) [reply bip-0127.mediawiki]: done, thanks

**Comment by stevenroose** [comment]
@luke-jr fixed the README

I don't currently have much time to work on this, but I think it's quite final apart from the CLTV issue. So can go in like this and I can amend with a fix for that if needed.

**Pull request merged**